[RECONSTRUCTION] [CLANG] `Loop not vectorized` warnings

Hello,

I am opening a separate issue for what we started discussing in https://github.com/cms-sw/cmssw/pull/41616.

CLANG IBs started showing some warnings reporting `loop not vectorized` as of the update to llvm16.
In concrete,  for the reconstruction module `RecoTracker/MkFitCore`:
```
>> Compiling  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/radix_sort.cc
/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/llvm/16.0.3-5d1904f3c7b5c53b7522ba0858baec44/bin/clang++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=110201 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DCMSSW_GIT_HASH='CMSSW_13_2_CLANG_X_2023-05-09-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_13_2_CLANG_X_2023-05-09-2300' -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/pcre/8.43-5dcc901acc02f624b22dd9840b2357e8/include -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/bz2lib/1.0.6-2c1f18484cb66c30aba7929f2be5e7d4/include -isystem/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/lcg/root/6.26.11-9720c7273eac1c892ddadde973cf1965/include -isystem/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/tbb/v2021.8.0-4e779f195a25a0aba119b27519937ba0/include -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/xz/5.2.5-83d0a00b575efd1701e07bedf7977343/include -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/zlib/1.2.11-3dfb2715f3608466b74431b80eb9d788/include -I/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/json/3.10.2-a6d86565b09ec3d0e02bf7b52c31bbfc/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -fuse-ld=bfd -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-c99-extensions -Wno-c++11-narrowing -D__STRICT_ANSI__ -Wno-unused-private-field -Wno-unknown-pragmas -Wno-unused-command-line-argument -Wno-unknown-warning-option -ftemplate-depth=512 -Wno-error=potentially-evaluated-expression -Wno-tautological-type-limit-compare -fsized-deallocation --gcc-toolchain=/pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161 -Ofast -fno-reciprocal-math -mrecip=none -DBOOST_DISABLE_ASSERTS -fopenmp-simd  -fPIC    -MMD -MF tmp/el8_amd64_gcc11/src/RecoTracker/MkFitCore/src/RecoTrackerMkFitCore/radix_sort.cc.d /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/radix_sort.cc -o tmp/el8_amd64_gcc11/src/RecoTracker/MkFitCore/src/RecoTrackerMkFitCore/radix_sort.cc.o
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/PropagationMPlex.cc:290:8: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   void helixAtRFromIterativeCCSFullJac(const MPlexLV& inPar,
       ^
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:248:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   void MkFinder::selectHitIndices(const LayerOfHits &layer_of_hits, const int N_proc) {
                 ^
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:248:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:693:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   void MkFinder::addBestHit(const LayerOfHits &layer_of_hits, const int N_proc, const FindingFoos &fnd_foos) {
                 ^
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:905:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   void MkFinder::findCandidates(const LayerOfHits &layer_of_hits,
                 ^
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:1151:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   void MkFinder::findCandidatesCloneEngine(const LayerOfHits &layer_of_hits,
                 ^
  /pool/condor/dir_43945/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/189e8911dd83c5e66aafe1fe240371ea/opt/cmssw/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_CLANG_X_2023-05-09-2300/src/RecoTracker/MkFitCore/src/MkFinder.cc:1151:18: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
 1 warning generated.
6 warnings generated.
```

I have seen in older PRs that sometimes this has been fixed by using a pragma statement, others inlining the function (which I believe it won't work here, but correct me if I am wrong) and in other cases the original authors took care of rearranging the code. Any hints on what can we do in those cases?

Many thanks! :)
Andrea.

FYI, @smuzaffar

--- a/RecoTracker/MkFitCore/src/MkFinder.cc
+++ b/RecoTracker/MkFitCore/src/MkFinder.cc
@@ -299,7 +299,11 @@ namespace mkfit {
     };
 
     if (L.is_barrel()) {
-      // Pull out the part of the loop that vectorizes
+      // Pull out the part of the loop that vectorizes with icc and gcc
+      // In llvm16 clang issues a warning that it can't vectorize
+      // the loop.  Unfortunately, there doesn't seem to be a
+      // pragma to suppress the warning, so we ifdef it out.  This
+      // should be rechecked if llvm vectorization improves.
 #if !defined(__clang__)
 #pragma omp simd
 #endif
@@ -341,7 +345,7 @@ namespace mkfit {
       //layer half-thikness for dphi spread calculation; only for very restrictive iters
       const float layerD = std::abs(L.layer_info()->zmax() - L.layer_info()->zmin()) * 0.5f *
                            (m_iteration_params->maxConsecHoles == 0 || m_iteration_params->maxHolesPerCand == 0);
-      // Pull out the part of the loop that vectorizes
+      // Pull out the part of the loop that vectorizes with icc and gcc
 #if !defined(__clang__)
 #pragma omp simd
 #endif
@@ -1210,7 +1214,7 @@ namespace mkfit {
                                      m_prop_config->finding_intra_layer_pflags,
                                      m_prop_config->finding_requires_propagation_to_hit_pos);
 
-       //#pragma omp simd  // DOES NOT VECTORIZE AS IT IS NOW
+      //#pragma omp simd  // DOES NOT VECTORIZE AS IT IS NOW
       for (int itrack = 0; itrack < N_proc; ++itrack) {
         // We can be in failed state from the initial propagation before selectHitIndices
         // and there hit_count for track is set to -1 and WSR state to Failed, handled below.

--- a/RecoTracker/MkFitCore/src/MkFinder.cc
+++ b/RecoTracker/MkFitCore/src/MkFinder.cc
@@ -300,7 +300,9 @@ namespace mkfit {
 
     if (L.is_barrel()) {
       // Pull out the part of the loop that vectorizes
+#if !defined(__clang__)
 #pragma omp simd
+#endif
       for (int itrack = 0; itrack < NN; ++itrack) {
         m_XHitSize[itrack] = 0;
 
@@ -340,7 +342,9 @@ namespace mkfit {
       const float layerD = std::abs(L.layer_info()->zmax() - L.layer_info()->zmin()) * 0.5f *
                            (m_iteration_params->maxConsecHoles == 0 || m_iteration_params->maxHolesPerCand == 0);
       // Pull out the part of the loop that vectorizes
+#if !defined(__clang__)
 #pragma omp simd
+#endif
       for (int itrack = 0; itrack < NN; ++itrack) {
         m_XHitSize[itrack] = 0;
 
@@ -716,7 +720,7 @@ namespace mkfit {
 
       mhp.reset();
 
-#pragma omp simd
+      //#pragma omp simd doesn't vectorize with current compilers
       for (int itrack = 0; itrack < N_proc; ++itrack) {
         if (hit_cnt < m_XHitSize[itrack]) {
           mhp.addInputAt(itrack, layer_of_hits.refHit(m_XHitArr.At(itrack, hit_cnt, 0)));
@@ -932,7 +936,7 @@ namespace mkfit {
 
       int charge_pcm[NN];
 
-#pragma omp simd
+      //#pragma omp simd doesn't vectorize with current compilers
       for (int itrack = 0; itrack < N_proc; ++itrack) {
         if (hit_cnt < m_XHitSize[itrack]) {
           const auto &hit = layer_of_hits.refHit(m_XHitArr.At(itrack, hit_cnt, 0));
@@ -1179,7 +1183,7 @@ namespace mkfit {
 
       int charge_pcm[NN];
 
-#pragma omp simd
+      //#pragma omp simd doesn't vectorize with current compilers
       for (int itrack = 0; itrack < N_proc; ++itrack) {
         if (hit_cnt < m_XHitSize[itrack]) {
           const auto &hit = layer_of_hits.refHit(m_XHitArr.At(itrack, hit_cnt, 0));
@@ -1206,7 +1210,7 @@ namespace mkfit {
                                      m_prop_config->finding_intra_layer_pflags,
                                      m_prop_config->finding_requires_propagation_to_hit_pos);
 
-#pragma omp simd  // DOES NOT VECTORIZE AS IT IS NOW
+       //#pragma omp simd  // DOES NOT VECTORIZE AS IT IS NOW
       for (int itrack = 0; itrack < N_proc; ++itrack) {
         // We can be in failed state from the initial propagation before selectHitIndices
         // and there hit_count for track is set to -1 and WSR state to Failed, handled below.

--- a/RecoTracker/MkFitCore/src/PropagationMPlex.cc
+++ b/RecoTracker/MkFitCore/src/PropagationMPlex.cc
@@ -297,6 +297,9 @@ namespace mkfit {
     MPlexLL errorPropTmp(0.f);   //initialize to zero
     MPlexLL errorPropSwap(0.f);  //initialize to zero
 
+    // loop does not vectorize with llvm16, and it issues a warning
+    // that apparently can't be suppressed with a pragma.  Needs to
+    // be rechecked if future llvm versions improve vectorization.
 #if !defined(__clang__)
 #pragma omp simd
 #endif

--- a/RecoTracker/MkFitCore/src/PropagationMPlex.cc
+++ b/RecoTracker/MkFitCore/src/PropagationMPlex.cc
@@ -297,7 +297,9 @@ namespace mkfit {
     MPlexLL errorPropTmp(0.f);   //initialize to zero
     MPlexLL errorPropSwap(0.f);  //initialize to zero
 
+#if !defined(__clang__)
 #pragma omp simd
+#endif
     for (int n = 0; n < NN; ++n) {
       //initialize erroProp to identity matrix
       errorProp(n, 0, 0) = 1.f;

